### PR TITLE
MPDX-8786 Add one time/recurring transfer mutations to Savings Fund report

### PR DESCRIPTION
**Pull request by kegrimes**
## Description

Since the GraphQL APIs -  `createTransfer`, `createRecurringTransfer`, `deleteRecurringTransfer`, and `updateRecurringTransfer` - have been merged,  we can connect these mutations to the front end. These have been tested and work as intended.

During this process, I discovered several issues that needed to be addressed. 

- Edit modal did not work with transfer history table because you can only edit the stop date once a recurring transfer has started (_Solution:_ Implement add/edit calendar icons)
- If a transfer is created in the future, there is no way for a user to view or edit that transfer until the first payment occurs (_Solution:_ add an upcoming transfers table)
- If a transfer were to fail, there is no way to see that (_Solution:_ add an information modal that breaks down recurring transfers)
- We still want to allow a transfer to happen with a negative account (_Solution:_ add a warning message in create transfer modal)

**Note:** We do not have tangible data for upcoming transfers yet.

Also, all test data has a `null` description, so I used `subCategory` name instead

## Checklist:

- [x] I have given my PR a title with the format "MPDX-(JIRA#) (summary sentence max 80 chars)"
- [x] I have applied the appropriate labels. (_Add the label "On Staging" to get the branch automatically merged into staging._)
- [x] I have requested a review from another person on the project

### CODE REVIEW

**Comment by kegrimes** [comment]
**Waiting for the final decision to be made regarding `deficit_limit`.**

Final tasks:
- Update modals based on decision
- Ensure time is updated only for the balance cards involved
- May need to remove edit modal due to constraints surrounding updating a recurring transfer

**Review by canac — CHANGES_REQUESTED** [review]
Thanks for creating these pages! The logic for filtering the transactions is pretty tricky, but it looks solid! I do have some thoughts on how it could be simplified. Sorry for all of the comments on that file.

This PR is also pretty large. It's definitely at the point where it could be good to think about ways to split it up into multiple PRs.

**Comment by canac** [comment]
I believe you were just implementing the design, but the dash makes it look like a negative amount to me. Could we just use a space or maybe a bullet • instead?

Screenshot 2025-09-30 at 12 11 02 PM

**Comment by canac** [comment]
Is there a way to add a label or title somewhere so that it's clear what date users are selecting for the add/edit end date modal?

Screenshot 2025-09-30 at 12 14 45 PM

File: src/components/Reports/SavingsFundTransfer/mockData.ts [hunk 1]
```diff
@@ -44,9 +66,13 @@ export interface Transactions {
...
   transfer: Transfer;
   recurringTransfer?: RecurringTransfer | null;
+  baseAmount: number;
+  failedCount?: number;
+  summarizedTransfers?: Map<string, Transactions> | null;
+  missingMonths?: DateTime[] | null;
 }
 
-export interface TransferHistory {
+export interface Transfers {
```
Comment (canac) [inline src/components/Reports/SavingsFundTransfer/mockData.ts:75]:
```suggestion
export interface Transfer {
```

File: src/hooks/useFilteredTransfers.ts [hunk 2]
```diff
@@ -1,33 +1,100 @@
...
-export function useFilteredTransfers(transfers: Transactions[]) {
+// It also identifies any missed transfers and includes them as separate transactions with a failed status.
+export function useFilteredTransfers(
+  transfers: Transactions[],
+  today?: DateTime,
+) {
   return useMemo(() => {
     const filtered: Transactions[] = [];
     const recurring = new Map<string, number>();
+    const seenMonths = new Map<number, Set<DateTime>>();
```
[the remark this answers is not in the document]
Comment (kegrimes) [reply src/hooks/useFilteredTransfers.ts:15]: I like this idea!

File: src/components/Reports/SavingsFundTransfer/Table/Row/createTableRow.tsx [hunk 3]
```diff
@@ -125,14 +189,75 @@ export const populateTransferHistoryRows = (
...
+            }}
+          >
+            <Icon
+              className="material-symbols-outlined"
+              sx={{
+                fontSize: 24,
+                color: 'cruGrayMedium',
+              }}
+            >
+              edit_calendar
```
[the remark this answers is not in the document]
Comment (kegrimes) [reply src/components/Reports/SavingsFundTransfer/Table/Row/createTableRow.tsx:226]: No, this text is so the Material symbol can render in the table. It is the name that is part of the font family Material Symbols Outlined. 

[Material Symbols](https://fonts.google.com/icons?selected=Material+Symbols+Outlined:edit_calendar:FILL@0;wght@400;GRAD@0;opsz@24&icon.size=24&icon.color=%239C9FA1&icon.query=calendar)

[the remark this answers is not in the document]
Comment (canac) [reply src/components/Reports/SavingsFundTransfer/Table/Row/createTableRow.tsx:226]: Is there a reason we can't use the Material Icons version?

`import EditCalendarIcon from '@mui/icons-material/EditCalendar';`

File: src/components/Reports/SavingsFundTransfer/mockData.ts (hunk 1 above)
Comment (kegrimes) [reply src/components/Reports/SavingsFundTransfer/mockData.ts:75]: I already have an interface named `Transfer` which matches the GraphQL query data.

File: src/components/Reports/SavingsFundTransfer/UpdatedAtContext/UpdateAtContext.tsx [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+  const context = React.useContext(UpdatedAtContext);
+  if (!context) {
+    throw new Error(
+      'Could not find UpdatedAtContext. Make sure that your component is inside <UpdatedAtProvider>',
+    );
+  }
+  return context;
+};
+
+export const UpdatedAtProvider: React.FC<
```
[the remark this answers is not in the document]
Comment (kegrimes) [reply src/components/Reports/SavingsFundTransfer/UpdatedAtContext/UpdateAtContext.tsx:20]: `BalanceCard` uses the context to get the updateAt value and sends it to the `useUpdatedAt` hook to get the label. `TransferModal` uses the context to set updatedAt when a create/update mutation is executed. I wasn't sure how else to go about it.

File: src/components/Reports/SavingsFundTransfer/TransfersPage/TransfersPage.tsx [hunk 5]
```diff
@@ -88,28 +93,42 @@ export const TransfersPage: React.FC<TransfersPageProps> = ({ title }) => {
...
-  const transferHistory: TransferHistory[] = filteredTransactions.map((tx) => {
-    const isRecurring = !!tx.recurringTransfer?.id;
+  const transferHistory: Transfers[] = filteredTransactions.map((tx) => {
+    const isRecurring = !!tx.recurringTransfer;
     const status = getStatusLabel(tx);
+    const shouldShowActions = () => {
+      return status !== StatusEnum.Pending && status !== StatusEnum.Ongoing;
+    };
 
     return {
```
Comment (canac) [inline src/components/Reports/SavingsFundTransfer/TransfersPage/TransfersPage.tsx:133]: May of these default values aren't needed because many of the `tx` fields aren't nullable.

File: src/components/Reports/SavingsFundTransfer/TransferModal/TransferModal.tsx [hunk 6]
```diff
@@ -177,217 +275,305 @@ export const TransferModal: React.FC<TransferModalProps> = ({
...
+                        item
+                        xs={12}
+                        sm={5.5}
+                        sx={{
+                          display: 'flex',
+                          justifyContent: 'flex-end',
+                          alignItems: 'center',
+                        }}
+                      >
+                        <FundInfoDisplay fund={fund as Fund} />
```
Comment (canac) [inline src/components/Reports/SavingsFundTransfer/TransferModal/TransferModal.tsx:392]:
```suggestion
                        <FundInfoDisplay fund={fund} />
```

File: src/components/Reports/SavingsFundTransfer/TransferModal/TransferModal.tsx [hunk 7]
```diff
@@ -177,217 +275,305 @@ export const TransferModal: React.FC<TransferModalProps> = ({
...
+                      </Grid>
+                      <Grid item xs={12} sm={1} sx={{ textAlign: 'center' }}>
+                        <East />
+                      </Grid>
+                      <Grid item xs={12} sm={5.5}>
+                        <FundInfoDisplay
+                          fund={
+                            funds.find((f) => f.fundType === transferTo) as Fund
```
Comment (canac) [inline src/components/Reports/SavingsFundTransfer/TransferModal/TransferModal.tsx:400]: If the fund can't be found, we probably want to not render `FundInfoDisplay`.

```suggestion
                            funds.find((f) => f.fundType === transferTo)
```

File: src/components/Reports/SavingsFundTransfer/TransferModal/TransferModal.tsx [hunk 8]
```diff
@@ -177,217 +275,305 @@ export const TransferModal: React.FC<TransferModalProps> = ({
...
+                        name="schedule"
+                        value={schedule}
+                        onChange={(_event, value) => {
+                          setFieldValue('schedule', value);
+                          if (value === ScheduleEnum.Monthly) {
+                            setFieldTouched('transferDate', true, false);
+                            validateField('transferDate');
+                          } else {
+                            validateField('transferDate');
+                          }
```
Comment (canac) [inline src/components/Reports/SavingsFundTransfer/TransferModal/TransferModal.tsx:426]:
```suggestion
                          }
                          validateField('transferDate');
```

File: src/components/Reports/SavingsFundTransfer/TransfersPage/TransfersPage.tsx [hunk 9]
```diff
@@ -123,10 +142,17 @@ export const TransfersPage: React.FC<TransfersPageProps> = ({ title }) => {
...
+      actions: shouldShowActions() === false ? 'edit-delete' : '',
+      recurringId: tx.recurringTransfer?.id || '',
+      baseAmount: tx.baseAmount || 0,
+      failedCount: tx.failedCount || 0,
+      summarizedTransfers: tx.summarizedTransfers,
+      missingMonths: tx.missingMonths,
     };
   });
 
+  const incoming = incomingTransfers;
```
Comment (canac) [inline src/components/Reports/SavingsFundTransfer/TransfersPage/TransfersPage.tsx:154]: Is the API for getting these transfers not ready yet?

File: src/components/Reports/SavingsFundTransfer/CustomEditCalendar/CustomEditCalendar.tsx [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+export const CustomEditCalendar: React.FC<CustomEditCalendarProps> = ({
+  open,
+  value,
+  onChange,
+  onClose,
+  onAccept,
+  minDate,
+  type,
+}) => {
+  const Toolbar: React.FC<DatePickerToolbarProps<DateTime>> = (props) => (
```
Comment (canac) [inline src/components/Reports/SavingsFundTransfer/CustomEditCalendar/CustomEditCalendar.tsx:119]: React doesn't like it when you declare a component within another component because it gets recreated during ever render. Do some research about the best way to handle this. It may be sufficient to wrap it in useMemo.

**Review by canac — COMMENTED** [review]
Thanks for going through all of my suggestions from before!

Do we still need to use mocked data for recurring transfers? The UI wasn't letting me edit recurring transfers because the ID was a mocked ID.

File: src/components/Reports/SavingsFundTransfer/UpdatedAtContext/UpdateAtContext.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (canac) [reply src/components/Reports/SavingsFundTransfer/UpdatedAtContext/UpdateAtContext.tsx]: Looks like you did this after I started reviewing. Thanks!

File: src/components/Reports/SavingsFundTransfer/DownloadTable/downloadTable.tsx [hunk 11]
```diff
@@ -33,7 +33,7 @@ export const createTable = (
       : transfer.status;
     const endDate =
       transfer.schedule === ScheduleEnum.Monthly && transfer.endDate
-        ? dateFormat(transfer.endDate, locale, { timezone: 'UTC' })
+        ? transfer.endDate.toFormat('MMM d, yyyy')
```
[the remark this answers is not in the document]
Comment (kegrimes) [reply src/components/Reports/SavingsFundTransfer/DownloadTable/downloadTable.tsx:36]: If I use `dateFormat` it will set the date to the day before. I was not sure how else to work around it besides using `.fromISO()` and format it myself.

File: src/components/Reports/SavingsFundTransfer/Table/Row/createTableRow.tsx [hunk 12]
```diff
@@ -60,11 +71,9 @@ export const populateTransferRows = (
     }
 
     return (
-      <Tooltip title={t('N/A') as string}>
-        <Typography variant="body2" noWrap>
-          {t('N/A') as string}
-        </Typography>
-      </Tooltip>
+      <Typography variant="body2" noWrap>
+        {t('N/A') as string}
```
[the remark this answers is not in the document]
Comment (kegrimes) [reply src/components/Reports/SavingsFundTransfer/Table/Row/createTableRow.tsx:75]: I left it because I had errors around that, but I just figured out I had the wrong import for `TFunction`!

File: src/components/Reports/SavingsFundTransfer/CustomEditCalendar/CustomEditCalendar.tsx (hunk 10 above)
Comment (kegrimes) [reply src/components/Reports/SavingsFundTransfer/CustomEditCalendar/CustomEditCalendar.tsx:119]: I put it in a context so I could pass `type` to the `CalendarTitle` component. I'm not sure if this is the correct solution, but it works well.

File: src/components/Reports/SavingsFundTransfer/Helper/filterTransfers.ts [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+
+    const transferCount = transferRow.amount / transferRow.baseAmount;
+
+    const monthsBetween = Math.floor(end.diff(start, 'months').months);
+    const expectedCount = Math.max(0, monthsBetween + 1);
+
+    if (transferCount < expectedCount) {
+      transferRow.failedCount = expectedCount - transferCount;
+
+      const missing: DateTime[] = [];
```
[the remark this answers is not in the document]
Comment (kegrimes) [reply src/components/Reports/SavingsFundTransfer/Helper/filterTransfers.ts:77]: Ohh, I see!

[the remark this answers is not in the document]
Comment (kegrimes) [reply src/components/Reports/SavingsFundTransfer/Helper/filterTransfers.ts:77]: Yes, that definitely simplifies the filtering! I will do that instead of calculating the actual count and expected count.
